target/ppc/translate/vsx-impl.c.inc

```diff
@@ -1446,19 +1446,15 @@ static void glue(gen_, name)(DisasContext *ctx)             \
 VSX_XXMRG(xxmrghw, 1)
 VSX_XXMRG(xxmrglw, 0)
 
-static void gen_xxsel(DisasContext *ctx)
+static bool trans_XXSEL(DisasContext *ctx, arg_XX4 *a)
 {
-    int rt = xT(ctx->opcode);
-    int ra = xA(ctx->opcode);
-    int rb = xB(ctx->opcode);
-    int rc = xC(ctx->opcode);
+    REQUIRE_INSNS_FLAGS2(ctx, VSX);
```

Comment:
Is ``REQUIRE_VSX`` below enough? In other words, is it possible for ``ctx->vsx_enabled`` to be true and ``PPC2_VSX`` to not be set?

If we do need both checks, what if we move the ``REQUIRE_INSNS_FLAGS2(ctx, VSX)`` inside ``REQUIRE_VSX``?

Reply:
> Is `REQUIRE_VSX` below enough? In other words, is it possible for `ctx->vsx_enabled` to be true and `PPC2_VSX` to not be set?

`ctx->vsx` is set from bit 40 of MSR. In older versions of PowerISA (v2.05 and bellow), this bit was reserved, so I think we cannot infer anything from it's value if the CPU does not have VSX.

> If we do need both checks, what if we move the `REQUIRE_INSNS_FLAGS2(ctx, VSX)` inside `REQUIRE_VSX`?

I guess we could move this check to `REQUIRE_VSX`, but we don't need to. If an instruction already checks for newer insns flags and we know that all CPUs that have that insns flag also have VSX, checking for PPC_VSX would be kind of redundant.

Reply:
Ok. I fear it might be kind of confusing for other people when to use (and when not to use) ``REQUIRE_INSNS_FLAGS2(ctx, VSX`` and ``REQUIRE_VSX``. This same problem also happens with Altivec. What do you think?

Would it be clearer if we called them ``REQUIRE_{VECTOR, VSX}_ENABLED`` or ``REQUIRE_MSR_{VECTOR, VSX}``, etc.? Or maybe I'm being overzealous here.

Reply:
There is a `PPC2_VSX207` too, that makes things even more confusing. Maybe the real problem is the number of insns flags. I'll dig some manuals and the mail list archive, but maybe we can drop `PPC_VSX` in favor of some `PPC2_ISA*` + `REQUIRE_VSX`.

Reply:
My understanding is that `PPC2_VSX` is a badly-named equivalent of `PPC_207` or `PPC_310` but for v2.06. AFAIK, by looking at the PowerISA book, v2.06 introduced VSX instructions, that's why it was decided to be called `PPC2_VSX`. However, VSX facility could still be disabled in this version or any other version, that's why `REQUIRE_VSX(ctx);` is necessary.

I agree it is confusing, my choice of action would be to suggest the renaming of this flag as an RFC.

Reply:
IIUC, `PPC2_VSX` was required because VSX is optional in PowerISA v2.06, so checking for the version (i.e.: adding a `PPC2_ISA206`) would not be enough. With the current behavior in QEMU, we'll throw an "Illegal Instruction Interrupt" when the facility is not present and a "VSX Unavailable Interrupt" when it's present but disabled.

Reply:
Oh, I see. I was wrong with what I assumed previously then. Thanks for the explanation. My understanding now is that `REQUIRE_INSNS_FLAGS2(ctx, VSX);` can throw an "Illegal Instruction Interrupt" if either is a version before v2.06 or if it is v2.06 but VSX is not present. Is that correct?

Reply:
`REQUIRE_INSNS_FLAGS2` itself only checks `ctx->insns_flags2`. You can check in `target/ppc/cpu_init.c` which CPUs have this flag (currently POWER{7,8,9,10}). For instance, e500mc is a PowerISA v2.06 implementation that lacks VSX, it's `insns_flags2` is initialized with `PPC2_BOOKE206` but not with `PPC2_VEX`.